### ISSUE TITLE
Combining destructuring with parameter properties

### ISSUE THREAD

**Issue by buzinas** [Microsoft/TypeScript#5326]
Today, we can take advantage of the parameter properties to reduce the boilerplate, e.g:

````javascript
class Person {
  constructor(public firstName : string, public lastName : number, public age : number) {

  }
}
````
Since 1.5, we can also use destructuring, e.g:

````javascript
class Person {
  firstName: string;
  lastName: string;
  age: number;
  constructor({firstName, lastName, age} : {firstName: string, lastName: string, age: number}) {
    this.firstName = firstName;
    this.lastName = lastName;
    this.age = age;
  }
}
````
I've tried in many ways to combine both features, but had no success. So:

 1. Is it possible to do that nowadays, and if yes, how?
 2. Could this be an improvement to a future TypeScript version, e.g:

````javascript
class Person {
  constructor({public firstName, public lastName, public age} : {firstName: string, lastName: string, age: number}) {

  }
}
````

**Comment by xLama** [comment]
I like this.

**Comment by DanielRosenwasser** [comment]
@buzinas I originally had this working in #1671, but didn't take it in (check out #1541). I think the issue was that it was relatively dense in semantics. I don't necessarily agree. Maybe someone else on the team can weigh in.

**Comment by danquirk** [comment]
Note that in those proposals the parameter property modifier applied to the entire binding pattern (simpler) as opposed to the original suggestion here which in theory supports different visibility modifiers per destructured element (not sure that level of specificity would be worth it).

**Comment by buzinas** [comment]
@danquirk For me, if it's easier for you to do the other way, I don't really care. In fact, that was my first try (e.g `public {firstName, lastName, age}`), and as soon as it didn't work, I tried to use on each property, and it didn't work too.

It would be great if we could support both (since not always we want to create properties for all the parameters, and when we want, it would be simpler to use `public`/`private` only once), but if it's easier to support only one approach, it will be great already.

Probably it's something that people will use more and more, since destructuring is awesome.

**Comment by robianmcd** [comment]
just ran into this. I think either approach would satisfy most use cases. Hope to see this in a future version.

**Comment by laurelnaiad** [comment]
The first thing I tried was:

```
  constructor(thisProps: { public myProperty: string, public myOtherProperty: number }) {
    // etc
  }
```

Something like this would be a nice-to-have.

**Comment by harangue** [comment]
Definitely +1 this. The alternative tends to be... bulky.

**Comment by garkin** [comment]
My suggestion:
```
class Person {
	constructor({
		public firstName:first as string, 
		public lastName:last as string, 
		publc age as number
	}) {}
}
const hero = new Person({first:"na", last:"labutenah", age:30};
```
This shouldn't interfere with any syntax now.

**Comment by albertywu** [comment]
:+1:

**Comment by hrundik** [comment]
:+1:

**Comment by aliatsis** [comment]
+1

**Comment by lekev** [comment]
+1

**Comment by andriilazebnyi** [comment]
+1

**Comment by yjaaidi** [comment]
👍

**Comment by TiuSh** [comment]
+1

**Comment by RyanCavanaugh** [comment]
Please use the GitHub reactions feature rather than standalone upvote comments. Thanks!

**Comment by appsforartists** [comment]
In an attempt to be more DRY using named args and strong types (until something like your proposal lands), I tried this:

```ts
interface ExampleArgs {
  firstArg: string;
  otherArg: number;
}

export default class Example implements ExampleArgs {
  firstArg;
  otherArg;

  constructor(kwargs:ExampleArgs) {
    return Object.assign(this, kwargs);
  }
}
```

but got `Member 'firstArg' implicitly has an 'any' type.` errors for every argument.  ☹️

**Comment by RyanCavanaugh** [comment]
Write this instead
```ts
interface ExampleArgs {
  firstArg: string;
  otherArg: number;
}

export default class Example {
  constructor(kwargs:ExampleArgs) {
    return Object.assign(this, kwargs);
  }
}
export interface Example extends ExampleArgs { }
```

**Comment by appsforartists** [comment]
Thanks.  I had to separate the exports from the declarations to make that work:

```typescript

interface ExampleArgs {
  firstArg: string;
  otherArg: number;
}

class Example {
  constructor(kwargs:ExampleArgs) {
    return Object.assign(this, kwargs);
  }
}

export default Example;
interface Example extends ExampleArgs { }
```

**Comment by markrendle** [comment]
This would be incredibly useful for hydrating class-based models from JSON.

As in

```typescript
export interface PersonDto {
    name?: string;
}

export class Person {
    constructor(public {name}: PersonDto = {}) {
    }
}
```

**Comment by yjaaidi** [comment]
Meanwhile we get this feature, here's the workaround:

```typescript
export class PersonSchema {
    firstName: string;
    lastName: string;
    email?: string; // Thanks to TypeScript 2, properties can be optional ;)
}

export class Person extends PersonSchema {

    constructor(args: PersonSchema = {}) {
        Object.assign(this, args);
    }

}

```

The side effect is that if `args` has extra fields they will be copied into your new object.

This will also be your copy constructor.

**Comment by RyanCavanaugh** [comment]
Accepting PRs to implement `constructor(public {name1, name2})` forms (i.e. no `public` / `private` etc *inside* the `{}`s).

**Comment by appsforartists** [comment]
@RyanCavanaugh, you're proposing that

```typescript
constructor(public { name1 = value1, name2 = value2 }: { name1: string, name2: string }) {}
```

would desugar to

```javascript
constructor({ name1 = value1, name2 = value2 }) {
  this.name1 = name1;
  this.name2 = name2;
}
```

right?

**Comment by dmoebius** [comment]
Why is this discussion only about constructors? Wouldn't this apply to other methods as well?!?

**Comment by yjaaidi** [comment]
It can't apply to other methods as only the constructor can declare
properties.

**Comment by AppLover69** [comment]
So, will this ever be implemented?

**Comment by andraaspar** [comment]
```TS
class Person {
    
    firstName: string
    lastName: string
    age: number
    
    constructor(_: Person) {
        Object.assign(this, _)
    }
}

class PersonWithEmail extends Person {
    
    email: string
    
    constructor(_: PersonWithEmail) {
        super(_)
    }
}

let p = new PersonWithEmail({
    firstName: '',
    lastName: '',
    age: 0,
    email: '',
})
```

**Comment by appsforartists** [comment]
@andraaspar I'd imagine you'd run into issues if there were methods on `Person`.  

```typescript
const model = {
  firstName: 'Andras',
}

class Person {
  firstName: string;
  doAThing() {
  }
}

const thisShouldError: Person = model;
```

**Comment by andraaspar** [comment]
@appsforartists Sure you would. Not a replacement for the real thing. But it's elegant for model objects, until it gets implemented.

**Comment by yjaaidi** [comment]
Hi, personally I'm using a `Schema` interface to avoid issues with methods etc... there's an example here: https://blog.wishtack.com/2017/05/06/angular-2-components-communication-using-reactive-stores/

The problem with `Object.assign` is that you might end up with dynamically added properties.
Example:
```
let data = {firstName: 'Foo', superfluousProperty: 'something'};
let person = new Person(data);
console.log(person['superfluousProperty']) // 'something'
```

It would be nice to be able to use `keyof` to generate a runtime array of class properties but this is not available for the moment. There are also some experimental approaches like this one https://github.com/kimamula/ts-transformer-keys.
Anyway, I'm sure, we'll get something nice soon :)

**Comment by chriszrc** [comment]
@yjaaidi this example fails, since some of the properties aren't optional. Just for people who are copying and pasting-

```
export class PersonSchema {
    firstName: string; //ERROR 
    lastName: string; //ERROR
    email?: string; // Thanks to TypeScript 2, properties can be optional ;)
}

export class Person extends PersonSchema {

    //ERROR firstName is not an optional property
    constructor(args: PersonSchema = {}) {
        Object.assign(this, args);
    }

}
```

however this works:

```
export class PersonSchema {
    firstName: string; 
    lastName: string; 
    email?: string; // Thanks to TypeScript 2, properties can be optional ;)
}

export class Person extends PersonSchema {

    constructor(args: PersonSchema = {firstName:"John",lastName:"Doe"}) {
        Object.assign(this, args);
    }

}
```

**Comment by HipsterZipster** [comment]
Really want this!

**Comment by phil-lgr** [comment]
Another take using Pick to make some props mandatory:

```ts
type PersonOptions = Partial<Person> & Pick<Person, 'firstName' | 'lastName'>;

class Person {
  firstName: string;
  lastName: string;
  age: number = 0;
  
  constructor(options: PersonOptions) {
      for (const key in options) {
        this[key] = options[key]
    }
  }
}

const person = new Person({
        firstName: 'first',
        lastName: 'last',
        age: 27
    }
)

const personeWithNoAge = new Person({
    firstName: 'first2',
    lastName: 'last2'
})
```

It seems I get full autocomplete!

**Comment by phil-lgr** [comment]
Extending my previous example with inheritance:

```ts
type SuperPersonOptions = PersonOptions & Partial<SuperPerson> & Pick<SuperPerson, 'superProperty'>;

class SuperPerson extends Person {

    superProperty: string;
    optionalSuperProperty: string = '';

    constructor(options: SuperPersonOptions) {
        super({
            firstName: options.firstName,
            lastName: options.lastName
        })
        for (const key in options) {
            this[key] = options[key]
        }
    }
}

const superPerson = new SuperPerson({
    superProperty: 'im super',
    firstName: 'superFirst',
    lastName: 'superLast'
})
```

**Comment by ryuuji3** [comment]
@phil-lgr don't you mean ConcretePerson?

**Comment by nukadelic** [comment]
@chriszrc we don't have to suffer typing all of that text, let machines do that for us: https://constructor.stackblitz.io

**Comment by krzkaczor** [comment]
@RyanCavanaugh are you still accepting PRs for this issue :D I would like to work on this.

**Comment by dvsoukup** [comment]
Just wanted to throw in my +1!  This would be super useful.  I'm guessing it's not much of a priority though being that this very issue was opened in 2015 :(

**Comment by ianldgs** [comment]
This would be extremely usefull for immutable classes:

```ts
class Immutable {
  constructor({
    public readonly fun: boolean,
    private readonly cool: boolean,
  }) { }

  get isCool(): boolean {
    return this.cool
  }
}
```

```ts
const immutable = new Immutable({fun: true, cool: true})
```

After spending some time with dart I started to value immutable objects and named constructors.
Really think it's a nice idea.

**Comment by yjaaidi** [comment]
Hi @ianldgs!

That's exactly the perfect pattern that needs this feature.
We also use it for "Entity" constructors as describe in our Angular guide here https://guide-angular.wishtack.io/typescript/duck-typing-patterns/entity-constructor (It's in french for the moment but google translate does a decent job).

**Comment by Google-Mac** [comment]
Just going to leave this here for anyone to use. I've noticed that many of these options above are pretty verbose, when Typescript does some of the work for you. For example, you do not need to declare the type of each parameter if the name matches the object property. I believe the type is inferred (correct me if I'm wrong, but that is what I've noticed).

```javascript
class Person
{
    _firstName: string;
    _lastName: string;
    _middleName?: string;

    constructor(data?: any) 
    {
        // exit if empty constructor
        if (typeof data === 'undefined') return;

        this.firstName = data.firstName || 'Joe';
        this.lastName = data.lastName || 'Schmoe';
        this.middleName = data.middleName;    
    }
}
```

**Comment by acchou** [comment]
Extending @andraaspar 's example to allow for methods:
```typescript
class Person {   
    firstName: string
    lastName: string
    age: number
    
    constructor(_: NonFunctionProperties<Person>) {
        Object.assign(this, _)
    }

    fullName() {
        return `${this.firstName} ${this.lastName}`;
    }
}
```
It also works well with default values for fields:
```typescript
class Person {   
    firstName = "";
    lastName = "";
    age = 0;
    
    constructor(_?: NonFunctionProperties<Person>) {
        Object.assign(this, _)
    }
}
```
`NonFunctionProperties<T>` is defined in the [2.8 release notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html) as:
```typescript
type NonFunctionPropertyNames<T> = { [K in keyof T]: T[K] extends Function ? never : K }[keyof T];
type NonFunctionProperties<T> = Pick<T, NonFunctionPropertyNames<T>>;
```
This also solves the problem with superfluous fields pointed out by @yjaaidi . It also works fine if some properties are optional; they also become optional in the constructor argument. On the downside, this approach doesn't play well with getters, which appear as non-functions to the type system.

**Comment by appsforartists** [comment]
@acchou 

Cool example, but a couple deficiencies: 

- All non-function properties are required.  TS will require you to provide all of them (or an arbitrary subset if you use `Partial`).  This could make API evolution more difficult, because it would make existing callsites invalid if you ever added a new property (unless you replaced this pattern with an explicit interface).
- It doesn't ignore `readonly`, which means it will require you to pass a value for any `readonly` parameter, even though they can't be written to.

**Comment by acchou** [comment]
@appsforartists If the properties are optional, then they won't be required in the constructor argument.   Invalidating callsites when non-optional properties are added is a feature, not a bug.

There is a case where only a subset of properties have a default value. In that case the example I've given requires either all of the properties, or none of them (as in the second example). That's a deficiency that seems harder to address right now because "having a default value" is not part of the property's type.

I don't see the issue with `readonly` as we're talking about constructor arguments. `readonly` properties need to be initialized too.

After playing a bit with the implementation it seems TypeScript will warn about properties potentially not being initialized in the constructor because it doesn't fully understand `Object.assign`. So this is mainly useful if you have default values for all properties, or use `!` or `?` on the property:

```
class Person {   
    firstName!: string;
    readonly lastName!: string;
    age?: number;
    
    constructor(_: NonFunctionProperties<Person>) {
        Object.assign(this, _);
    }

    fullName() {
        return `${this.firstName} ${this.lastName}`;
    }
}
```

**Comment by ab18556** [comment]
+1 on this feature too.

I read all the comments, and I don't see what's the real show stopper with this syntax:

`class Person {
  constructor({ public firstName, public lastName, private age }: { firstName: string, lastName: string, age?: number }) {}
}`

The visibility of the properties is obvious and the requirement of a variable can be based on the one set in the destructured object's type. The property requirement would be a little bit obscured, but it should't be game breaking since this feature would be so awesome to have as is.

I really don't like the Object.assign solution. It can fit as a temporary solution for some developers, but the intent is obscure.

And writing this.firstName = firstName; is way more frustrating when you know that a better solution is already implemented for non-destructured params.

Without this, I feel I have to chose between destructuring and in-param declaration if I want to keep my code readable. This is heartbreaking.

It's been 2 years since this issue was created. Let's not take another 2 years to decide if it should be done or not ;)

**Comment by buzinas** [comment]
> It's been 2 years since this issue was created. 

3 years already. In the meantime I stopped using TypeScript at all, but I still receive notifications about this thread I created. I don't stop watching because I'm pretty curious to know when this will be solved.

**Comment by igpeev** [comment]
How about extending the original proposition a little bit to handle also "default" values, like so:
```
class Person {
  constructor ({
    public name: string = 'John',
    public age: number = 21,
    public isEnrolled = true,      // <-- implicit type 
    private ssn: string,
  }) { }
}
```
and used like:
`let p = new Person({name: 'Peter', ssn: '555-777-44'});`

vs traditional:
let p = new Person('Peter', undefined, undefined, '555-777-44');     // <-- params meaning not so obvious

**Comment by appsforartists** [comment]
@igpeev I think we'd all like to be able to put the key and type in the braces.  Unfortunately, that conflicts with JavaScript's existing renaming syntax, so I doubt a solution that includes it will make much progress.  (TypeScript is supposed to be a strict superset of JavaScript, not a fork.)

To someone who doesn't speak TypeScript, it looks like you're renaming `name` and `ssn` to `string` and `age` to `number`.  How would the compiler know when to choose types vs. when to rename?  How would someone rename a local property in your proposal?

**Comment by igpeev** [comment]
'public'/'private' keywords could so nicely solve the dilemma, had they not been reserved for the 'future'. 
P.S. Does anyone really use this renaming, would happily sacrifice it to have this constructor sugar

**Comment by ianldgs** [comment]
@appsforartists You are absolutely right.
@igpeev I don't think that typescript maintainers would sacrifice that.

But one workaround would be to prepend `this:` in the constructor:

dart` 

```ts
class Immutable {
  constructor(this: {
    public readonly fun: boolean,
    private readonly cool: boolean,
  }) { }

  get isCool(): boolean {
    return this.cool
  }
}
```

**Comment by appsforartists** [comment]
:thinking: Clever!

There's already a meaning for [this as an argument](https://www.typescriptlang.org/docs/handbook/functions.html), but TypeScript [bans that use in constructors](http://www.typescriptlang.org/play/#src=class%20Test%20%7B%0A%20%20constructor(this%3A%20void)%20%7B%0A%0A%20%20%7D%0A%7D), so this could be a solution.

**Comment by ab18556** [comment]
> (TypeScript is supposed to be a strict superset of JavaScript, not a fork.)
Maybe TypeScript should be forked then ;)

Seriously, I get that TypeScript maintainers want to keep typescript as close as possible to EcmaScript and I think it is the right choice, but it would be nice if we could create and share plugins to enable such features.

This isn't something that can be fixed via a language service plugin https://github.com/Microsoft/TypeScript/wiki/Writing-a-Language-Service-Plugin

Anyone know a way we could do it?

**Comment by appsforartists** [comment]
Not to derail this thread, but one of the great things about TypeScript (vs Babel) is that it's curated.  Playing with Babel plugins is fun, but every library doesn't need its own dialect of the language.

**Comment by ab18556** [comment]
I am just saying that if TypeScript must be kept as close as possible to EcmaScript, having a way to create plugins would be a great way to augment it in a non-disruptive way.

> That said, you can run Babel on top of TypeScript

I am not sure I like this idea. So far, TypeScript allowed me to avoid Babel and I am happy I don't have this extra layer of complexity into my environment.

> Not to derail this thread

Well this thread has been open for more than three years. Maybe it needs to be derailed a little bit in order to get it somewhere...

**Comment by micheleangioni** [comment]
It's strange TypeScript still does not support something like this. 
A great amount of code could be avoided easily.

**Comment by jack-guy** [comment]
I attempted an implementation for this feature a few months ago. I got stuck on some test cases, but this isn't as hard to do as you might imagine.

People in this thread who are passionate about this feature should give implementing it a go - the TypeScript team is accepting PRs for a reason! You'll probably get further than I did!

**Comment by rodrigolive** [comment]
Current constructor parameters can make class declarations very hard to read since properties show up in two different sections (and indentations) of a class, and mixing up some proposal ideas and standard JS here we could end up with default values in 4 different places:

```typescript
// this is not "real" code
// but what would go through the TS beginner's mind
// as they think of where properties and their default values should go...

class FooBar {
      foo:number = 1;
      constructor(this: { public foo: number = 2 } = { foo: 3 }) {
            this.foo = 4;
      }
}
```

No matter what restrictions we impose, any proposals for implementing destructuring in the constructor declaration will only lead to cluttered, hard to read code. Also, destructuring doesn't allow for transforming or coercion of initialization data.

There are 2 ways to go about it that would be more elegant.

One would be to add sugar to this nice, and currently supported, construct inspired by React's `this.props`:

```typescript
type Properties<T> = Pick<
    T,
    { [K in keyof T]: T[K] extends Function ? never : K }[keyof T]
>;

class FooBar {
    foo: number = this.props.foo * 10;
    bar?: number = this.props.bar;
    qaz?: number = this.props.qaz == null ? 33 : this.props.qaz;
    private distance : number;

    constructor(private props: Properties<FooBar>) {}
}

const foobar : FooBar = new FooBar({ foo: 11, bar: 22 });
```

I like the fact that:

- constructed properties are clearly indicated by the `= this.props.<property>` idiom
- constructor properties can be pre-processed and coerced during assignment, ie. `= this.props.foo * 10`
- `this.props` remains around as a reference to the original properties, useful for object cloning or re-instantiation.
- clearer and less global than using `Object.assign()`

And you can also easily add additional constructor-only parameters:

```typescript
class FooBar {
    // ...
    bar?: number = this.props.baz * 10;
    constructor(private props: Properties<FooBar> & { baz? : number }) {}
}
```

If you prefer, put default values in the constructor instead of in the property initializer:

```typescript
class FooBar {
     foo : number = this.props.foo;
     constructor(private props : Properties<FooBar> = { foo: 123 }) {}
}
```

It would be great to avoid the boilerplate `type Properties ...` (or `NonFunctionProperties`) every time. So any sugar in that direction would be great, ie a `Properties`, `Props`, `NonFunction` or similar core TS sugar that is concise and clear.

Also support for TC39 nullish coalescing `??` operator would make property initializers simpler:

```typescript
class FooBar {
     public foo : number = this.props.foo ?? 123;
     constructor(private props);  // also don't require a constructor implementation
}
``` 

Now, if new syntax is to be introduced to properties, then it should be in the direction of convenient **constructed properties**, inspired by languages like Perl6:

Here's a proposal of a new `construct` keyword that would tell TS that the property will be built from the constructor object argument *without the need for a constructor declaration/implementation*:

```typescript
class FooBar {
    construct private foo: number = this.foo * 10; // this.foo has value input by the user
    construct public bar?: number;
    private distance : number;
}

new FooBar({ foo: 11, bar: 22 });
new FooBar({ bar: 22 });  // Error: missing 'foo'
new FooBar({ foo: 11, distance: 33 });  // Error: 'distance' is not constructed
```

The `constructor()` method could still be available, but with a few limitations:

```typescript
class FooBar {
      construct foo : number;
      bar:number;
      constructor(mystr: string) { } // Error: mismatched constructs
      constructor(private props) { }  // ok, props type is { foo: number }
      constructor(public props) { }  // Error TS2502, props is referenced directly or indirectly on its own type annotation
      constructor(props : any) { }  // Error: mismatched constructs
      constructor(props) {
            this.foo = this.foo || 99; // ok 
            this.bar = this.foo * 10;  // ok to keep building
      }
}
```
This would give developers great convenience and flexibility while being clear about intentions.

**Comment by oppianmatt** [comment]
I found this works for me, let's me inject into constructors, or I can destructure and create:

```

export class Person{
  constructor(
    readonly firstName: string,
    readonly lastName: string,
    readonly email?: string,
  ) {}

  static new(args: Person) {
    Object.assign(this, args)
  }
}

const person1 = new Person('first', 'last')
const person2 = new Person('first', 'last', 'foo@example.com')
const person3 = Person.new({ firstName: 'first', lastName:'last'})
const person4 = Person.new({ firstName: 'first', lastName: 'last', email: 'foo2@example.com'})
```

**Comment by markfinlabs** [comment]
I have a class for incoming JSON objects with two known fields, ```id``` and ```owner```.

It currently looks like this:
```typescript
class DataItem {
  id: string;
  owner: string;
  constructor({
    id,
    owner,
    ...otherFields
  } : {
    id: string;
    owner: string;
    [others: string]: string | number | boolean;
  }) {
    this.id = id;
    this.owner = owner;
    Object.assign(this, otherFields);
  }
}
```

Would be great to have following simplified syntax that avoids specifying type of ```id``` and ```owner``` twice:

```typescript
class DataItem {
  constructor({
    public id,
    public owner,
    ...otherFields
  } : {
    id: string;
    owner: string;
    [others: string]: string | number | boolean;
  }) {
    Object.assign(this, otherFields);
  }
}
```

**Comment by Soaku** [comment]
The above solution is great, it's compatible with ECMAScript and works well with current syntax. Unlike the `public {...}: {...}` destructing property, it also offers more possibilities, though that syntax would be nice to see too. Is anyone willing to implement this? Maybe I'd look into it this week.

**Comment by LeanHer94** [comment]
Hi there! Any news on this?

**Comment by casamia918** [comment]
Woooow. It's really amazing. I can't believe that ts still not implement this 'basic' language feature. Many js developers preferred named parameters at class constructor to make more clear codes. Auto assigning at constructor is great idea. Then, why do you guys not think that developers also want to use the auto assigning feature in named parameter way? It's been FIVE years! Many great ideas are already mentioned above.

**Comment by ianldgs** [comment]
This is probably not going to be implemented anytime soon. I think none of the proposals made everyone feel good. For now, I'm using a workaround based on a comment someone sent here previously but I can't find it to give the credits.
My main goal was to have immutable classes like dart, so `private`/`protected` doesn't really matter for me.

```ts
interface IUser {
    readonly id: number
    readonly name: string
}

// merge the interface with the class
interface User extends IUser {}

class User {
    constructor(props: IUser) {
        Object.assign(this, props)
    }

    public copyWith(props: Partial<IUser>) {
        // spread operator only copies enumerable properties, so no need to worry about methods
        return new User({ ...this, ...props })
    }

    public sayHello() {
        console.log(this.name, 'says hello')
    }
}

const oldUser = new User({ id: 1, name: 'Foo' })
const newUser = oldUser.copyWith({ name: 'Bar' })

oldUser.sayHello() // "Foo says hello"
newUser.sayHello() // "Bar says hello"

oldUser.id = 2; // error: readonly
newUser.id = 2; // error: readonly
```

**Comment by SrBrahma** [comment]
@ianldgs that's very good, with the available ways. Shortening it to the objective, we have:

```
interface IUser {
    id: number;
    name: string;
}

// merge the interface with the class
interface User extends IUser {}

class User {
    constructor(props: IUser) {
        Object.assign(this, props);
    }
}
```

-------------------------------------------------------------

It should be shorted to something like this:
```
interface IUser {
    id: number;
    name: string;
}

class User uses IUser {
    constructor(props: IUser) {
        Object.assign(this, props);
    }
}
```

Where **uses** could be a new word, or extends / implements could work this way.

This Object.assign still bothers me a bit, but I think it's possible to get used to it.

--------------------------------------

Just to remember people the problem it solves:

```
interface IUser {
    id: number;
    name: string;
}

class User implements IUser {
    id: number;
    name: string;
    constructor(props: IUser) {
        this.id = props.id
        this.name = props.name;
    }
}
```

It requires the programmer basically writing 3 times the same intention. I **really** hate doing this. "Oh, you want to quickly add a new information to the class? You must write it 3 times".

Writing 2 variables is already a pain, and when I must write many, it makes me want to scratch my brain >:/

**Comment by SrBrahma** [comment]
My uses solution, that shortens
```
interface User extends IUser {}
class User {...}
```
into 
```
class User uses IUser {...}
```
looks like a good idea, if isn't possible to make _implements_ automatically adding the interface variables into the class, and leaves _extends_ with its requirements of having a _super()_.

"Now", it would only need to make Typescript to watch if all the required variables are initialized, as the
```
interface User extends IUser {}
class User {
```
(or as I am calling, _uses_)
removes the checking of initializations. Also, would require that Typescript watches the Object.assign results.

**Comment by cmawhorter** [comment]
one thing to remember is that interface is an internal typescript thing only and blindly doing Object.assign(this, props) is dangerous on the server and can fall down in the real world on the server/browser in weird and unexpected ways. consider:

```ts
interface IUser {
    readonly id: number
    readonly name: string
}

interface User extends IUser {}

class User {
    private admin = false;

    constructor(props: IUser) {
        Object.assign(this, props)
    }
}
```

compiles to this js:

```js
"use strict";
class User {
    constructor(props) {
        this.admin = false;
        Object.assign(this, props);
    }
}
```

and then let's say on the server you're doing:

```js
const someUserInput = { id: 'user_123', name: 'bob', admin: true } // <== er337 hax0ring
const user = new User(someUserInput)
console.log('admin should be false, but: ', user.admin);
```

you could probably even do something like `new User({ constructor: { prototype: {} } })` and replace the entire prototype maybe? 

tl;dr object.assign is a poor workaround unless you're also validating/filtering input but at that point you might as well _type things three times_

**Comment by ianldgs** [comment]
@cmawhorter you're absolutelly right. `Object.assign(this, any)` is extremely dangerous if you're not validating input.
I'm my case, I always validate input for extra props, even in projects that I don't need this workaround.

**Comment by SrBrahma** [comment]
> ```ts
> class User {
>   private admin = false;
> 
>   constructor(data: Partial<User>) {
>     public id: string = data.id;
>     public name: string = data.name
>     protected featureFlag = data.featureFlag ?? false;
>   }
> }
> ```

@cmawhorter

This is really good and really could be implemented. I would improve with this:

```ts
class User {
  private admin = false;

  constructor(data: Partial<User>) {
    public id: string;
    public name: string;
    protected featureFlag = data.featureFlag ?? false;
  }
}
```

Without the initializer and with the Partial<$Class> parameter, it would implicitly mean that they would receive their value from the Parcial argument.

**Comment by casamia918** [comment]
How about this. Let's create new type keyword `Props`.

```typescript
class User1 {

  public name: string = prop1.; // shorthand for prop binding
  public age: number = prop1. * 10; // computed prop binding

  constructor(prop1: Props) {}

}

// In this way, typescript compiler know that type Props of User1 class should conform below type

type PropsInUser1 = {
  name: string;
  age: number;
}

```



```typescript
class User2Wrong {

  public name: string = prop2.; // shorthand for prop binding
  public nickname: string = prop2.nick // Compile Error. ts compiler can't infer field name 'nick' is included in prop2

  constructor(prop2: Props) {}

}

class User2Correct {

  public name: string = prop2.; // shorthand for prop binding
  public nickname: string = prop2.nick // Now ts compiler can know field name 'nick' is included in prop2 because of intersectioning of {nick:string}

  constructor(prop2: Props & {nick: string}) {}

}


```


```typescript
class User3 {

  public name: string;
  public age: number;

  constructor(prop3: Props) {
     // ordinary prop assignment
    this.name = prop3.name;
    this.age = prop3.age;
  }

}



```

**Comment by mattdarveniza** [comment]
Introducing a new keyword as common as `Props` would break about 95% of codebases that already use that at least once as a `type Props` or `interface Props`. I don't see the TS team being keen on introducing any new keywords for a feature that at best only reduces a small amount of boilerplate.

**Comment by casamia918** [comment]
@mattdarveniza Code conflict can be resolved by using compilerOption. I can't agree with your last sentense. This is not small part of boilerplate code. When using named parameter, developer has to type 3 times repeatedly to bind ctor parameter to class property. New keyword is worth to do.

**Comment by jake-jester-nih** [comment]
Something I've been playing around with: (Not in potential PR, just writing what I'd like to see)

interface IFooPublic {
  thin: string,
  large: number,
}

interface IFooPrivate {
  mine: any,
  
}

class Foo implements IFoo {
  constructor( public ... IFooPublic, private ...IFooPrivate ) {}
}

The idea is that you are destructuring the interface and adding it to the available properties of the class, while also providing the non-class interfaces for pushing data around. /shrug

**Comment by GusBuonv** [comment]
As I saw no push back against it and a good showing of support for it, I'd like to resurface and bump @markm77's suggestion on this. To repeat the proposed syntax:

```typescript
interface Data {
  foo: string
  bar: string
}

class A {
  // new syntax simultaneously declares properties with access and mutability specifiers and initializes them from argument properties
  constructor({
    readonly public foo,
    readonly public bar,
  }: Data) {}
}
```

To me this suggestion has the following advantages:

* It solves the original problem.
* It is an easy-to-understand extension of an existing feature (JavaScript's argument unpacking).
* It cuts down on boilerplate.
* It reduces the potential for errors in constructors (e.g. accidentally assigning `data.foo` to `this.bar`, unsafe & unnecessary reliance on `Object.assign`, etc).
* It facilitates a basic design pattern: classes wrapping interfaces with functionality.
* It does not rely on exotic or unfamiliar extensions to the language.

I would like to see this option get more discussion, so here are some questions that might get this moving. Is there anything that makes this impossible or overly burdensome to implement from a technical standpoint? Is there anything about the syntax that makes it unclear or prone to causing errors? Does this syntax fall short of facilitating the feature at the root of this thread?

**Comment by sandersn** [comment]
I put this back into help-wanted because it feels like a hole in the language. [This comment](https://github.com/microsoft/TypeScript/issues/5326#issuecomment-260505168) summarises the design decision at the time -- `public/private` go outside the destructuring.

**Comment by ianldgs** [comment]
> I put this back into help-wanted because it feels like a hole in the language. [This comment](https://github.com/microsoft/TypeScript/issues/5326#issuecomment-260505168) summarises the design decision at the time -- `public/private` go outside the destructuring.

What about `readonly`? What about having some fields `public` and some `private`? How would it look like?
This design is old and seems a bit vague to me. There are a lot of better ideas in other comments.

**Comment by charles-allen** [comment]
> I put this back into help-wanted because it feels like a hole in the language. [This comment](https://github.com/microsoft/TypeScript/issues/5326#issuecomment-260505168) summarises the design decision at the time -- `public/private` go outside the destructuring.

Please correct me if I'm wrong, but I read that comment to simply mean that the _first version_ should be the simplest MVP. It does not proclude later versions extending the syntax to increase flexibility. It is _not_ making a design decision to oppose additional syntax (later).

**Comment by ianldgs** [comment]
> > I put this back into help-wanted because it feels like a hole in the language. [This comment](https://github.com/microsoft/TypeScript/issues/5326#issuecomment-260505168) summarises the design decision at the time -- `public/private` go outside the destructuring.
> 
> Please correct me if I'm wrong, but I read that comment to simply mean that the _first version_ should be the simplest MVP. It does not proclude later versions extending the syntax to increase flexibility. It is _not_ making a design decision to oppose additional syntax (later).

Makes kinda sense, but IMO this is a small feature enough that creating an MVP to extend it further later feels unnecessary and would actually make it harder to extend later, leading to a possible breaking change to actually extend it.

**Comment by Antoniossss** [comment]
If I recall correctly, Dart has something similar. I think it is called named params, and it can be bound directly in constructor. 

It can be checked via DartPad. How it works

class Person {
  int ?age; // ts age?:number;
  String ?name; //ts name?:string;
  Person({this.age, this.name});
}

void main() {
  new Person(age:5);
  new Person(name:"Sebastian");
  new Person(age:5,name:"Sebastian");
}

This reduces boilerplate by 1 level by removing assignment statements (better than nothing)
```

In TS it could be like 

```
class Person {
  private age? : number;
  private name? : string;
  constructor({this.age, this.name});
}

new Person({name:'sebastian'}); 
```

What do you think?

**Comment by KutnerUri** [comment]
this is great, and I'm looking forward to #44912.

Why only pre-destructure  syntax, though?
```ts
constructor(public { id, secretHash }: ...) {}
// vs
constructor({ public id, private secretHash }: ...) {}
```

**Comment by zzyyxxww** [comment]
I find this combination of declaration merging and `Object.assign` to be the easiest way to deal with this at the moment:
```typescript
interface ClsAttrs {
  name: string;
}

interface Cls extends ClsAttrs {}
class Cls {
  constructor(attrs: ClsAttrs) {
    Object.assign(this, attrs)
  }
}
```

**Comment by luke-robertson** [comment]
Bump

**Comment by luke-robertson** [comment]
Still want this in 2022

**Comment by larssn** [comment]
I'd love to have a Dart-like syntax for this.

We'd be able to do:

```typescript
class Test {
  constructor({
    private param1: string,
    protected param2: number,
    public param3?: string // optional
  }); // <-- Might as well make the constructor branckets optional too!
}

const test = new Test({
  param1: 'hello',
  param2: 1337,
  param3: 'world',
});

console.log(test.param3); // -> 'world'
```

**Comment by ianldgs** [comment]
@larssn that is what I proposed here:
https://github.com/microsoft/TypeScript/issues/5326#issuecomment-404910771
However, someone rightfully pointed out that it conflicts with existing JS syntax: 
https://github.com/microsoft/TypeScript/issues/5326#issuecomment-440905301
As an alternative, `this` could be added in front to differentiate:
https://github.com/microsoft/TypeScript/issues/5326#issuecomment-441789833

But to be honest, I haven't used classes in JS/TS in a while. And not planning to use it again anytime soon.
To reflect: what is the difference between immutable pojos + functions vs immutable classes?

```ts
interface Obj {
  readonly foo: string;
  readonly bar: string;
}

function foobar(obj: Obj) {
  return obj.foo + obj.bar;
}
```

```ts
class Obj {
  public readonly foo: string;
  public readonly bar: string;
  
  // whatever constructor
  
  foobar() {
    return this.foo + this.bar;
  }
}
```

For me, there isn't much, except that with immutable pojos + functions I don't have to worry about `this` context.

**Comment by larssn** [comment]
@ianldgs There's should be no difference between them, it's just semantics.

Anyway, consider my post to just be a big fat thumbs up for your post then! 🙂 
But I don't understand why renaming properties (#5326) would be a hindrance - tooling shouldn't influence the spec, unless I'm misunderstanding something.

**Comment by ianldgs** [comment]
Hm, actually, thinking back, the presence of the access modifier should already indicate it's not renaming

`constructor({ a: b })` is it renaming `a` to `b` or saying that `a` is of type `b`?
But
`constructor({ public a: b })` is clear that it's declaring a `public` property `a` of type `b`.

unless I'm also missing something

**Comment by Akxe** [comment]
@sandersn what part does need to be discarded. Maybe if we knew all that needed to be decided beforehand we could create a poll to vote on the best solution. I know it has not have been done here before, but it could provide great result :)

**Comment by ayoreis** [comment]
This is my I workaround with default values for simple classes that don't have methods.

```ts
class Attribute {
  name!: string
  value!: string
  
  constructor(args: Partial<Attribute>) {
    Object.assign(this, { name: '', value: '' }, args)
  }
}
```

**Comment by REPTILEHAUS** [comment]
Is using `Object.assign(this, data)` still the DRY recommended approach to de-structure ?

**Comment by Roosteridk** [comment]
Still want this in 2023

**Comment by weoreference** [comment]
Hello is there a roadmap for this feature? Would greatly improve the ergonomics of typescript!

**Comment by a-tarasyuk** [comment]
@luke-robertson @Roosteridk @weoreference  https://github.com/microsoft/TypeScript/pull/44912#issuecomment-1090809726

**Comment by Akxe** [comment]
How about a syntax with a new keyword? The keyword `deconstructing` can only be used before a `constructor` class method.

This is an example of the possible syntax:
```typescript
class Test {
  deconstructing constructor(
    private readonly param1: string;
    protected param2?: number;
    public param3: string;
    param4 = param3;
  ) {
    // param2 is optional and yet can be followed by non-optional parameters
    // param4 is local to the constructor only
  }
}
```
Some nuances:
* The parameter list is not wrapped in braces (`{ }`)
* Every line ends with `;` and not `,` to minify mystification with normal constructor

The proposed syntax:
* Is not easily confused with normal constructor syntax yet similar enough to be easily readable
* Does define types for the constructor
* Does allow for all types of parameters, including local to the constructor

Downsides:
* Does not allow for the renaming of parameters
* Does not (currently) handle overloading

And this is the produced code/typescript
```typescript
class Test {
  private readonly param1: string;
  protected param2?: number;
  public param3: string;

  constructor(options: {
    param1: string;
    param2?: number;
    param3: string;
    param4?: string
  }) {
    this.param1 = options.param1;
    this.param2 = options.param2;
    this.param3 = options.param3;
    var param4 = options.param4 ?? param2;
    // param4 is local to the constructor only
  }
}
```

**Comment by Tomaszal** [comment]
I really like the solution proposed by @rodrigolive in https://github.com/Microsoft/TypeScript/issues/5326#issuecomment-455155622. Most of it can even be achieved already by adding branded types to their example (I'm sure this can be improved, suggestions are welcome):

```ts
declare const tag: unique symbol;

type ConstructProp<T> = T & { [tag]: typeof tag };

type ConstructedProps<Class> = Pick<
  Class,
  {
    [Key in keyof Class]: Required<Class>[Key] extends ConstructProp<unknown>
      ? Key
      : never;
  }[keyof Class]
>;

type ConstructorProps<Class> = {
  [Key in keyof ConstructedProps<Class>]: Required<
    ConstructedProps<Class>
  >[Key] extends ConstructProp<infer PropType>
    ? PropType
    : never;
};
```

Then, class props can be defined like so (perhaps someone better at TypeScript can suggest a way to avoid casting to satisfy branded types here, or at least a way to delegate that to a generic function):

```ts
class FooBar {
  foo: ConstructProp<number> = this.props.foo as this['foo'];
  bar: ConstructProp<number> = (this.props.bar * 7) as this['bar'];
  baz?: ConstructProp<number> = this.props.baz as this['baz'];
  qaz?: number;

  constructor(private readonly props: ConstructorProps<FooBar>) {}
}

const foobar = new FooBar({ foo: 3, bar: 5 });
```

Unfortunately, [`useDefineForClassFields`](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#the-usedefineforclassfields-flag-and-the-declare-property-modifier) has to be set to `false` for this to work, which is not ideal, as it strays from the recommended defaults. I don't think a similar solution is currently possible with `useDefineForClassFields` set to `true`.

**Comment by Akxe** [comment]
@Tomaszal This is a nice type-safe workaround until a proper solution is in place, but it would break for any generic class, where generic used to be defined by parameter in constructor.

Also a big negative is that you cannot mimic defining class properties in constructor. This results in looking thought whole class declaration for any constructor props that might be present. Also private properties would not be properly found, since they are not visible to other function/types. And lastly class extending would break since the properties of parent would be requested in children too.

**Comment by awlayton** [comment]
My current workaround is to create a class decorator that will destructure the first constructor argument "automagically"

```typescript
/**
 * Class decorator for automagically destructuring the constructor argument
 */
function destructure<T, R extends [...unknown[]]>(
  Klass: Constructor<T, R]>,
  _context?: unknown,
): Class<T, [Partial<T>]> {
  const properties = Object.getOwnPropertyNames(
    new Klass({}),
  ) as unknown as Array<keyof T>;
  return class
    extends // @ts-expect-error classes as variables nonsense
    Klass
  {
    constructor(argument: Partial<T> = {}) {
      const a = [];
      for (const property of properties) {
        a.push(argument[property]);
      }

      super(...a);
    }
  } as unknown as Class<T, [Partial<T>]>;
}
```

With I can use parameter properties in the class like so:
```typescript
export @destructure class Foo {
  constructor(
    public foo: string,
    public bar: number,
  ) {}
}
```

But then I will be able to call the decorated constructor with an object:
```typescript
  const foo = new Foo({bar: 1, foo: 'hi'});
```

**Comment by kibiz0r** [comment]
@awlayton Having to `new` up an instance in order to interrogate the fields creates a dangerous situation for constructors that have side-effects.

If I have a singleton pattern like:

```ts
export @destructure class Foo {
  static instance: Foo;
  constructor(
    public foo: string,
    public bar: number,
  ) {
    Foo.instance ??= this;
  }
}
```

...now `Foo.instance` is an instance which never got a `foo` or `bar` argument.

**Comment by awlayton** [comment]
Yes that is true @kibiz0r. My solution is only helpful for very specific cases. Language support for it would be much better.

**Comment by kibiz0r** [comment]
Worth noting that the `this.props` pattern suggested by @rodrigolive has suffered some bit rot since they posted it.

A class like this will break if you're using the ES2022 syntax for class fields, which is the default for target ES2022 or higher, including ESNext. For this to work, you'd need to set [`useDefineForClassFields`](https://www.typescriptlang.org/tsconfig#useDefineForClassFields) to `false` in your tsconfig.

```ts
class Foo {
  constructor(private props: { foo: number }) {}
  private foo = this.props.foo;
}
```

**Comment by HasanMothaffar** [comment]
We still want this in 2024

**Comment by a-tarasyuk** [comment]
@HasanMothaffar https://github.com/microsoft/TypeScript/pull/44912#issuecomment-1090809726

**Comment by Akxe** [comment]
@a-tarasyuk Where is the discussion happening? I want to follow that thread.

**Comment by a-tarasyuk** [comment]
@Akxe you can find the details on why the PR aiming to resolve this issue was closed here https://github.com/microsoft/TypeScript/pull/44912#issuecomment-1090809726. Alternatively, you can ask @RyanCavanaugh or @DanielRosenwasser for more detailed explanations...

**Comment by Akxe** [comment]
@a-tarasyuk The comment above is in reaction to your [comment in #44912]. (https://github.com/microsoft/TypeScript/pull/44912#issuecomment-1091222081). My question is where the mentioned discussion is happening. This thread is almost dead yet there seems to be a lot of interest for this specific feature in the community. If you think that the other leading members should add to this discussion, please welcome them here, or we could even schedule some kind of "meeting" to discuss this.

**Comment by a-tarasyuk** [comment]
@Akxe ok. I apologize if I provided misleading information that confused you...

**Comment by sukima** [comment]
What are the blockers to this suggestion? How can we get Parameter Properties support with Named Arguments?

**Comment by k41n** [comment]
👍

**Comment by RoyTinker** [comment]
It's strange to look for and find an open issue based on something I'm currently working on, and then see that I voted for it like 8 years ago

**Comment by Akxe** [comment]
> What are the blockers to this suggestion? How can we get Parameter Properties support with Named Arguments?

The biggest blocker is that this is purely a TS problem. Native JS have no problem with destructuring of this kind since they do not have the public/private/private keywords. That means that there would have to be a proposal backed by someone in the TS inner circle. 

------

**My opinion**: (Take it as you want, correct me if I am wrong...)
Currently, the TS team does not want to add anything to the TS that will not be added to JS. They do not want to get into the same troubles as they got into with decorators, or import mismatched behaviours. They also do not want to make new keywords to prevent possible collisions with new future JS keywords.

------

Here is an overview of what was proposed (chronologically):
| Author        | Tiny example                                      | My comment                                | Link                                                                        |
|---------------|---------------------------------------------------|-------------------------------------------|-----------------------------------------------------------------------------|
| RyanCavanaugh | constructor(public {name1, name2})                | Very important persona                    | https://github.com/microsoft/TypeScript/issues/5326#issuecomment-260505168  |
| ianldgs       | constructor({ public readonly fun: boolean })     | 25+ likes; public/private per property    | https://github.com/microsoft/TypeScript/issues/5326#issuecomment-404910771  |
| casamia918    | public a = props.b; constructor(props1: Props)    | You need to see the comment to understand | https://github.com/microsoft/TypeScript/issues/5326#issuecomment-666321301  |
| Antoniossss   | private name?: string; constructor({this.name})   | Dart syntax in TS                         | https://github.com/microsoft/TypeScript/issues/5326#issuecomment-900943650  |
| Akxe (me)     | deconstructing constructor(private name: string;) | "deconstructing" keyword + new syntax     | https://github.com/microsoft/TypeScript/issues/5326#issuecomment-1526656240 |

------

If I forgot anyone, I am sorry; it is over 100 messages with very similar-looking responses a lot of the time.

**Comment by ianldgs** [comment]
@Akxe worth noting that my initial proposal linked has a conflict with existing JS syntax, as pointed out in this comment:
https://github.com/microsoft/TypeScript/issues/5326#issuecomment-440905301

My revised proposal:
https://github.com/microsoft/TypeScript/issues/5326#issuecomment-441789833

This one is also interesting:
https://github.com/microsoft/TypeScript/issues/5326#issuecomment-455155622

Edit: my current workaround:
https://github.com/microsoft/TypeScript/issues/5326#issuecomment-592058988

**Comment by RyanCavanaugh** [comment]
> Currently, the TS team does not want to add anything to the TS that will not be added to JS. They do not want to get into the same troubles as they got into with decorators, or import mismatched behaviours. They also do not want to make new keywords to prevent possible collisions with new future JS keywords.

This is basically correct. The odds we'd add destructuring to parameter properties is very, very low at this point. A proposal for something like this really needs to be taken to the TC39 committee in some form or another (not sure how since visibility is not a JS concept, but since parameter properties don't work with `#` anyway, you could imagine just forgetting about `protected`).

**Comment by Akxe** [comment]
If public/protected/private properties have not been implemented, would you be in favour of implementing them at all nowadays? If so, I feel like the reason not to extend their functionality is counterintuitive. 

Anyhow, given the fact that TC39 will never see private/protected/public, it is an opportunity to add something to the language that will never have the problem of conflicting with TC39's future intentions with JS.

**Comment by RyanCavanaugh** [comment]
> If public/protected/private properties have not been implemented, would you be in favour of implementing them at all nowadays?

The visibility modifiers on property declarations, yes.

Parameter properties, absolutely not.

> I feel like the reason not to extend their functionality is counterintuitive.

The relevant idiom here is "if you're in a hole, the first step is to stop digging"
